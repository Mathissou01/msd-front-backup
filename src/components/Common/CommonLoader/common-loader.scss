@use "src/styles/01-settings/settings.colors" as colors;
@use "src/styles/02-tools/tools.z-index" as index;
@use "src/styles/02-tools/tools.skeleton" as skeleton;
@use "src/styles/02-tools/tools.typography" as typo;

.c-CommonLoader {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;

  &_absolute {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }

  &_grow {
    flex-grow: 1;
  }

  &_skeleton {
    &::after {
      @include skeleton.loading;

      position: absolute;
      z-index: index.z("background");
      width: 100%;
      height: 100%;
      border-radius: 10px;
      content: "";
    }
  }

  &_cover {
    &::before {
      position: absolute;
      z-index: index.z("content-cover");
      width: 100%;
      height: 100%;
      content: "";
      pointer-events: none;
    }
  }

  &__Errors {
    display: flex;
    flex-grow: 1;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
  }
}
