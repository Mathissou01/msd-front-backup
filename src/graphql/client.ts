import {
  ApolloClient,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";

let client: ApolloClient<NormalizedCacheObject>;

if (process.env.NEXT_PUBLIC_MOCK === "true") {
  /* With .env variable NEXT_PUBLIC_MOCK=true, return mock graphql client with fake response data */
  client = (await import("./mockedClient")).default(1000);
} else {
  client = await new ApolloClient({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
    cache: new InMemoryCache(),
  });
}

export default client;
