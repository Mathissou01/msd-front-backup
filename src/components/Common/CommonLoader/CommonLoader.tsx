import classNames from "classnames";
import { ApolloError } from "@apollo/client";
import { ReactNode, useEffect, useRef, useState } from "react";
import CommonSpinner from "../CommonSpinner/CommonSpinner";
import "./common-loader.scss";

interface ICommonLoaderProps {
  isLoading: boolean;
  isShowingContent?: boolean;
  isCover?: boolean;
  isFlexGrow?: boolean;
  hasSpinner?: boolean;
  hasSkeleton?: boolean;
  hasDelay?: boolean;
  minHeight?: number;
  errors?: Array<ApolloError | undefined>;
  children: ReactNode;
}

export default function CommonLoader({
  isLoading,
  isShowingContent = false,
  isCover = true,
  isFlexGrow = true,
  hasSpinner = true,
  hasSkeleton = false,
  hasDelay = false,
  minHeight,
  errors,
  children,
}: ICommonLoaderProps) {
  const hasErrors = errors?.some((error) => error);

  const [isShowingLoader, setIsShowingLoader] = useState(!hasDelay);
  const spinnerTimerRef = useRef<NodeJS.Timeout>();
  useEffect(() => {
    if (isLoading) {
      if (!isShowingLoader) {
        spinnerTimerRef.current = setTimeout(() => {
          setIsShowingLoader(true);
        }, 200);
      }
    } else {
      clearTimeout(spinnerTimerRef.current);
      setIsShowingLoader(false);
    }
  }, [isLoading, isShowingLoader]);

  const wrapperClassnames = classNames("c-CommonLoader", {
    "c-CommonLoader_absolute": isShowingContent,
    "c-CommonLoader_cover": isCover,
    "c-CommonLoader_grow": isFlexGrow,
    "c-CommonLoader_skeleton": hasSkeleton,
  });

  return (hasDelay ? isShowingLoader : isLoading) ? (
    <>
      <div
        className={wrapperClassnames}
        style={minHeight ? { minHeight: `${minHeight}px` } : {}}
      >
        {hasSpinner && <CommonSpinner />}
      </div>
      {isShowingContent && children}
    </>
  ) : hasErrors ? (
    <div className="c-CommonLoader__Errors">
      {errors?.map((error, index) => {
        if (error && error.message) {
          return <span key={`error_${index}`}>{error?.message}</span>;
        }
      })}
    </div>
  ) : (
    <>{children}</>
  );
}
