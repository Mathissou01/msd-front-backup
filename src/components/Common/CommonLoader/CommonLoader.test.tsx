import { render } from "@testing-library/react";
import CommonLoader from "./CommonLoader";
import { ApolloError } from "@apollo/client";

describe("CommonLoader", () => {
  it("renders loading state", () => {
    const { container } = render(
      <CommonLoader isLoading={true} hasDelay={false}>
        <span>loaded content</span>
      </CommonLoader>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders loaded state", () => {
    const { container } = render(
      <CommonLoader isLoading={false} hasDelay={false}>
        <span>loaded content</span>
      </CommonLoader>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders error state", () => {
    const error = new ApolloError({ errorMessage: "error message" });
    const { container } = render(
      <CommonLoader isLoading={false} hasDelay={false} errors={[error]}>
        <span>loaded content</span>
      </CommonLoader>,
    );
    expect(container).toMatchSnapshot();
  });
});
