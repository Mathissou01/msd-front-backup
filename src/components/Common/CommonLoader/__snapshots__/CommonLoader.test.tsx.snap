// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonLoader renders error state 1`] = `
<div>
  <div
    class="c-CommonLoader__Errors"
  >
    <span>
      error message
    </span>
  </div>
</div>
`;

exports[`CommonLoader renders loaded state 1`] = `
<div>
  <span>
    loaded content
  </span>
</div>
`;

exports[`CommonLoader renders loading state 1`] = `
<div>
  <div
    class="c-CommonLoader c-CommonLoader_cover c-CommonLoader_grow"
  >
    <div
      class="c-CommonSpinner__Wrapper"
    >
      <div
        class="c-CommonSpinner"
        data-testid="common-spinner"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;
